rust/src/operations/update.rs

```diff
@@ -0,0 +1,936 @@
+        self
+    }
+
+    /// Perform an additonal update expression during the operaton
+    pub fn with_update<S: Into<Column>, E: Into<Expression>>(
+        mut self,
+        column: S,
+        expression: E,
+    ) -> Self {
+        self.updates.insert(column.into(), expression.into());
+        self
+    }
+
+    /// The Datafusion session state to use
+    pub fn with_session_state(mut self, state: SessionState) -> Self {
+        self.state = Some(state);
+        self
+    }
+
+    /// Additional metadata to be added to commit info
+    pub fn with_metadata(
+        mut self,
+        metadata: impl IntoIterator<Item = (String, serde_json::Value)>,
+    ) -> Self {
+        self.app_metadata = Some(Map::from_iter(metadata));
+        self
+    }
+
+    /// Writer properties passed to parquet writer for when fiiles are rewritten
+    pub fn with_writer_properties(mut self, writer_properties: WriterProperties) -> Self {
+        self.writer_properties = Some(writer_properties);
+        self
+    }
+
+    /// Specify the cast options to use when casting columns that do not match the table's schema.
+    pub fn with_cast_options(mut self, cast_options: CastOptions) -> Self {
+        self.cast_options = cast_options;
+        self
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+async fn execute(
+    predicate: Option<Expression>,
+    updates: HashMap<Column, Expression>,
+    object_store: ObjectStoreRef,
+    snapshot: &DeltaTableState,
+    state: SessionState,
+    writer_properties: Option<WriterProperties>,
+    app_metadata: Option<Map<String, Value>>,
+    cast_options: CastOptions,
+) -> DeltaResult<((Vec<Action>, i64), UpdateMetrics)> {
+    // Validate the predicate and update expressions
+    //
+    // If the predicate is not set then all files needs to be updated.
+    // else if only contains partitions columns then perform in memory-scan
+    // otherwise scan files for records that statisfy the predicate
+    //
+    // For files that were identified, scan for record that match the predicate
+    // and perform update operations, and then commit add and remove actions to
+    // the log
+
+    let exec_start = Instant::now();
+    let mut metrics = UpdateMetrics::default();
+    let mut version = snapshot.version();
+
+    if updates.is_empty() {
+        return Ok(((Vec::new(), version), metrics));
+    }
+
+    let predicate = match predicate {
+        Some(predicate) => match predicate {
+            Expression::DataFusion(expr) => Some(expr),
+            Expression::String(s) => Some(snapshot.parse_predicate_expression(s)?),
+        },
+        None => None,
+    };
+
+    let mut _updates = HashMap::new();
+    for (key, expr) in updates {
+        let expr = match expr {
+            Expression::DataFusion(e) => e,
+            Expression::String(s) => snapshot.parse_predicate_expression(s)?,
+        };
+        _updates.insert(key, expr);
+    }
+    let updates = _updates;
```

Comment:
```suggestion
    let updates = updates
        .into_iter()
        .map(|(key, expr)| {
            let expr = match expr {
                Expression::DataFusion(e) => e,
                Expression::String(s) => snapshot.parse_predicate_expression(s)?,
            };
            Ok((key, expr))
        })
        .collect::<Result<HashMap<_, _>>, _>()?;
```

I think this might also work a be a little bit more idiomatic?